### ISSUE TITLE
Agda fails to compile code involving @0

### ISSUE THREAD

**Issue by nad** [agda/agda#4168]
Agda fails to compile the following code properly:
```agda
module Bug where

record Erased (A : Set) : Set where
  constructor [_]
  field
    @0 erased : A

open Erased

record R (M : Set → Set) : Set₁ where
  field
    f : (A B : Set) → M A → (A → M B) → M B

works : R Erased
R.f works _ _ x g .erased = erased (g (erased x))

fails : R Erased
R.f fails _ _ x g = [ erased (g (erased x)) ]
```
GHC complains:
```
[…]/MAlonzo/Code/Bug.hs:31:50: error:
    • Variable not in scope: d10
    • Perhaps you meant one of these: ‘d16’ (line 16), ‘d30’ (line 22)
   |
31 | d38 = coe C47 (coe (\ v0 v1 v2 v3 -> coe v3 (coe d10 erased)))
   |                                                  ^^^
```
The generated Haskell code:
```haskell
{-# LANGUAGE EmptyDataDecls, EmptyCase, ExistentialQuantification,
             ScopedTypeVariables, NoMonomorphismRestriction, RankNTypes,
             PatternSynonyms #-}
module MAlonzo.Code.Bug where

import MAlonzo.RTE (coe, erased, AgdaAny, addInt, subInt, mulInt,
                    quotInt, remInt, geqInt, ltInt, eqInt, eqFloat, add64, sub64,
                    mul64, quot64, rem64, lt64, eq64, word64FromNat, word64ToNat)
import qualified MAlonzo.RTE
import qualified Data.Text

name4 = "Bug.Erased"
d4 a0 = ()
data T4 = C12
name16 = "Bug.R"
d16 a0 = ()
newtype T16
  = C47 (() -> () -> AgdaAny -> (AgdaAny -> AgdaAny) -> AgdaAny)
name30 = "Bug.R.f"
d30 ::
  T16 -> () -> () -> AgdaAny -> (AgdaAny -> AgdaAny) -> AgdaAny
d30 v0
  = case coe v0 of
      C47 v1 -> coe v1
      _ -> MAlonzo.RTE.mazUnreachableError
name32 = "Bug.works"
d32 :: T16
d32 = coe C47 erased
name38 = "Bug.fails"
d38 :: T16
d38 = coe C47 (coe (\ v0 v1 v2 v3 -> coe v3 (coe d10 erased)))
```
I'm guessing that `d10` corresponds to the projection `erased`, which is not compiled:
```
Not compiling Erased.erased.
```

**Comment by andreasabel** [comment]
Looking at debug output, it seems that compiler and type checker are not to be blamed, but an invalid __eta-contraction__, which turns the definition into:
```agda
fails : R Erased
R.f fails _ _ x g = g (erased x)
```

**Comment by nad** [comment]
The following code exhibits a similar problem:
```agda
open import Agda.Builtin.Equality
open import Agda.Builtin.Sigma

P : (A B : Set) → (A → B) → Set
P A B f = (y : B) → Σ A (λ x → f x ≡ y)

record R (A B : Set) : Set where
  field
    to      : A → B
    from    : B → A
    from-to : (x : A) → from (to x) ≡ x

p : (A B : Set) (r : R A B) → P A B (R.to r)
p _ _ r y = R.from r y , to-from y
  where
  postulate
    to-from : ∀ x → R.to r (R.from r x) ≡ x

record Erased (A : Set) : Set where
  constructor box
  field
    @0 erased : A

p′ : (A : Set) → P (Erased A) (Erased (Erased A)) (box {A = Erased A})
p′ A = p _ _ (record
  { to      = box {A = _}
  ; from    = λ { (box (box x)) → box {A = A} x }
  ; from-to = λ x → refl {A = Erased A} {x = _}
  })
```
Error message:
```
[…]/MAlonzo/Code/Bug.hs:52:34: error:
    • Variable not in scope: d72
    • Perhaps you meant one of these:
        ‘MAlonzo.Code.Agda.Builtin.Equality.d12’ (imported from MAlonzo.Code.Agda.Builtin.Equality),
        ‘d22’ (line 18), ‘d42’ (line 34)
   |
52 | du78 = coe du50 (coe C179 erased d72)
   |                                  ^^^
```
Here `d72` corresponds to `erased`.

Agda generates the following code:
```haskell
du78 :: T66 -> MAlonzo.Code.Agda.Builtin.Sigma.T14
du78 = coe du50 (coe C179 erased d72)
```
The code is compiled without complaints if the first underscore in `p′` is replaced by `Erased A`, or if the second underscore is replaced by `Erased (Erased A)`, or if the third underscore is replaced by `Erased A`, or if the fourth underscore is replaced by `x`. In either of these cases Agda generates the following code:
```haskell
du78 :: T66 -> MAlonzo.Code.Agda.Builtin.Sigma.T14
du78 = coe du50 (coe C179 erased erased)
```

**Comment by nad** [comment]
It seems as if Agda replaces the pattern-matching lambda with `erased`. Here are some fragments of the debug print-out (for `-vtc:20`):
```
new meta _36 : (A : Set) → Erased (_A_33 (A = A)) → _A_33 (A = A)
for postponed typechecking problem λ { (box (box x))
                                         → box {A = A} x
                                     } :?
                                   Erased _A_33 → _A_33
{ etaExpandMeta _37
postponing eta-expansion of meta variable _37 which is blocked by _A_33
}
adding constraint _37 = _36 : Erased _A_33 → _A_33
adding constraint [] _36
                       := λ { (box (box x)) → box {A = A} x } :? Erased _A_33 → _A_33
}
Checking named arg from-to = λ x → refl {A = Erased A} {x = _} :
                   (x : _A_33) → _37 (box x) ≡ x
```
```
solving _37 := λ A → Erased.erased
```
```
waking boxed constraint Erased.erased = λ { (box (box x)) → box x }
                          : Erased (Erased A) → Erased A
```

**Comment by andreasabel** [comment]
At first, the term in `from`, extended lambda ` λ { (box (box x)) → box {A = A} x }`, is a postponed type checking problem assigned to meta, say: 
```
  ?f : Box (Box A) -> Box A
```  
Field `from-to` has type `(x : Box A) -> ?f (box x) == x` and value `\ x -> refl` which means that Agda attempts the conversion:
```
  x : Box A |- ?f (box x) = x : Box A
```
This apparently solves `?f = unbox`:
```
compareTerm (not syntactically equal) _37 (box x) = x : Box A
attempting shortcut _37 (box x) := x
passed occursCheck
type of meta = (A : Set) → Box (Box A) → Box A
solving _37 := λ A → Box.unbox
shortcut successful
```
`inverseSubst` turns `x : Box A |- ?f (box x) = x` into:
```
y : Box (Box A) |- ?f y = unbox y
```
This is more a reprise of #4185 (rather than of an eta-contraction issue).

**Comment by andreasabel** [comment]
For now, we do not uncurry metas over records that have an erased field any more.  
Maybe this could be relaxed in case the record argument to the meta is erased, like in this artificial example:
```agda
{-# OPTIONS -v tc.meta.assign:25 #-}

open import Agda.Builtin.Equality

record Box (@0 A : Set) : Set where
  constructor box
  field @0 unbox : A

mutual
  X : {@0 A : Set} (@0 b : Box A) → Box (Box A)
  X = _

  spec : ∀{@0 A : Set} {@0 x : A}
       → X (box x) ≡ box (box x)
  spec = refl
```
Here we should be able to massage the constraint into the form:
```
@0 y : Box A |- X y ≡ box (box (unbox y))
```
The rationale is that if the record pattern is an erased argument, then all pattern variables will appear in erased context only.  Thus, it is save to use the projections from the record value there.
